plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -252,23 +320,22 @@ private void addScanTaskToSplits(TupleDomain<IcebergColumnHandle> dynamicFilterP
             List<org.apache.iceberg.DeleteFile> fullyAppliedDeletes = tableHandle.getEnforcedPredicate().isAll() ? scanTask.deletes() : ImmutableList.of();
             scannedFiles.add(new DataFileWithDeleteFiles(scanTask.file(), fullyAppliedDeletes));
         }
-        if (!fileTasksIterator.hasNext()) {
+        if (fileAwareScanTask.isLastSplitOfParentFile()) {
```

Comment:
Instead of adding isLastSplitOfParentFile/parentFileHasAnyDeletion and checking them from here, can we make `tasks()` method stop when "limit is reached" ? It seems to have all the required info to do this.

Reply:
The reason the limit is checked here, is that `tasks()` doesn't know which splits will actually be read and which will be filtered out by the dynamicFilter when the `getNextBatch()` is called. `tasks()` can only filter scanTasks  based on static information (in `filterScanTask()`), but dynamicFilter might still filter some of these scanTasks out later on

Reply:
If limit is pushed into table scan, we don't want to apply dynamic filtering at all as that is just wasted work. In general we don't push predicate through limit, so a dynamic filter won't be present in the plan if a limit is present over the scan. Even if some future optimization were to push both dynamic filter and limit into the scan, we would want to ignore the dynamic filter in that case.